src/gui/qgscheckablecombobox.cpp

```diff
@@ -0,0 +1,205 @@
+/***************************************************************************
+                              qgscheckablecombobox.cpp
+                              ------------------------
+  begin                : March 21, 2017
+  copyright            : (C) 2017 by Alexander Bruy
+  email                : alexander dot bruy at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgscheckablecombobox.h"
+
+#include <QLineEdit>
+#include <QAbstractItemView>
+
+
+QgsCheckableItemModel::QgsCheckableItemModel( QObject *parent )
+  : QStandardItemModel( 0, 1, parent )
+{
+}
+
+Qt::ItemFlags QgsCheckableItemModel::flags( const QModelIndex &index ) const
+{
+  return QStandardItemModel::flags( index ) | Qt::ItemIsUserCheckable;
+}
+
+QVariant QgsCheckableItemModel::data( const QModelIndex &index, int role ) const
+{
+  QVariant value = QStandardItemModel::data( index, role );
+
+  if ( index.isValid() && role == Qt::CheckStateRole && !value.isValid() )
+  {
+    value = Qt::Unchecked;
+  }
+
+  return value;
+}
+
+bool QgsCheckableItemModel::setData( const QModelIndex &index, const QVariant &value, int role )
+{
+  bool ok = QStandardItemModel::setData( index, value, role );
+
+  if ( ok && role == Qt::CheckStateRole )
+  {
+    emit dataChanged( index, index );
+    emit itemCheckStateChanged();
+  }
+
+  return ok;
+}
+
+
+QgsCheckBoxDelegate::QgsCheckBoxDelegate( QObject *parent )
+  : QStyledItemDelegate( parent )
+{
+}
+
+void QgsCheckBoxDelegate::paint( QPainter *painter, const QStyleOptionViewItem &option, const QModelIndex &index ) const
+{
+  QStyleOptionViewItem &nonConstOpt = const_cast<QStyleOptionViewItem &>( option );
+  nonConstOpt.showDecorationSelected = false;
+  QStyledItemDelegate::paint( painter, nonConstOpt, index );
+}
+
+
+QgsCheckableComboBox::QgsCheckableComboBox( QWidget *parent )
+  : QComboBox( parent )
+  , mSeparator( ", " )
+  , mDefaultText( "" )
+{
+  setModel( new QgsCheckableItemModel( this ) );
+  setItemDelegate( new QgsCheckBoxDelegate( this ) );
+
+  QLineEdit *lineEdit = new QLineEdit( this );
```

Comment:
I was thinking about adding context menu to the dropdown with two options: "select all" and "clear selection"